File: Validações/Algoritmos de validação/Python/validacao_heap_sort.py
import os

def ajustar_heap(X, tamanho, i):
    """
    Função auxiliar para manter a propriedade de max-heap.
    """
    maior = i
    esq = 2 * i + 1
    dir = 2 * i + 2

    if esq < tamanho and X[esq] > X[maior]:
        maior = esq

    if dir < tamanho and X[dir] > X[maior]:
        maior = dir

    if maior != i:
        X[i], X[maior] = X[maior], X[i]
        ajustar_heap(X, tamanho, maior)

def heapsort(X):
    """
    Função principal que aplica o algoritmo Heap Sort em uma lista.
    """
    tamanho = len(X)

    for i in range(tamanho // 2 - 1, -1, -1):
        ajustar_heap(X, tamanho, i)

    for i in range(tamanho - 1, 0, -1):
        X[i], X[0] = X[0], X[i]  
        ajustar_heap(X, i, 0)   

def ler_vetor(nome_arquivo):
    """Lê um vetor de um arquivo, com tratamento de erro."""
    try:
        with open(nome_arquivo, "r") as arquivo:
            vetor = [int(linha.strip()) for linha in arquivo]
            return vetor
    except FileNotFoundError:
        print(f"Erro: Arquivo de entrada não encontrado em {nome_arquivo}")
        return None
    except Exception as e:
        print(f"Ocorreu um erro ao ler o arquivo {nome_arquivo}: {e}")
        return None

def escrever_vetor(nome_arquivo, vetor):
    """Escreve um vetor (lista) em um arquivo de texto."""
    try:
        with open(nome_arquivo, "w") as arquivo:
            for numero in vetor:
                arquivo.write(f"{numero}\n")
    except Exception as e:
        print(f"Ocorreu um erro ao escrever no arquivo {nome_arquivo}: {e}")

def main():
    """
    Função principal que orquestra a leitura, ordenação com Heap Sort
    e escrita de 100 vetores.
    """
    caminho_entrada_base = os.path.join("..", "..", "..", "Vetores", "tamanho_100")
    caminho_saida_base = os.path.join("..", "..", "Vetores ordenados", "Python", "Heap sort")

    os.makedirs(caminho_saida_base, exist_ok=True)

    print("Iniciando processo de ordenacao em lote com Heap Sort (Python)...")

    for i in range(1, 101):
        nome_arquivo_entrada = os.path.join(caminho_entrada_base, f"vetor_100_{i}.txt")
        nome_arquivo_saida = os.path.join(caminho_saida_base, f"vetor_100_{i}.txt")

        vetor = ler_vetor(nome_arquivo_entrada)

        if vetor is not None:
            heapsort(vetor)
            escrever_vetor(nome_arquivo_saida, vetor)
            print(f"Arquivo vetor_100_{i}.txt processado com sucesso.")

    print("Processo concluido!")


if __name__ == "__main__":
    main()

File: Validações/Algoritmos de validação/C++/validacao_heap_sort.cpp
#include <iostream>
#include <vector>
#include <fstream>
#include <string>
#include <filesystem> 
#include <algorithm>  

void ajustarHeap(std::vector<int>& X, int tamanho, int i) {
    int maior = i;
    int esq = 2 * i + 1;
    int dir = 2 * i + 2;

    if (esq < tamanho && X[esq] > X[maior]) {
        maior = esq;
    }

    if (dir < tamanho && X[dir] > X[maior]) {
        maior = dir;
    }

    if (maior != i) {
        std::swap(X[i], X[maior]);
        ajustarHeap(X, tamanho, maior);
    }
}

void heapsort(std::vector<int>& X) {
    int n = X.size();

    for (int i = n / 2 - 1; i >= 0; i--) {
        ajustarHeap(X, n, i);
    }

    for (int i = n - 1; i > 0; i--) {
        std::swap(X[0], X[i]);
        ajustarHeap(X, i, 0);
    }
}

std::vector<int> lerVetorDeArquivo(const std::string& nomeDoArquivo) {
    std::ifstream arquivo(nomeDoArquivo);
    if (!arquivo.is_open()) {
        std::cerr << "Erro ao abrir o arquivo de entrada: " << nomeDoArquivo << std::endl;
        return {};
    }
    std::vector<int> vetor;
    int numero;
    while (arquivo >> numero) {
        vetor.push_back(numero);
    }
    return vetor;
}

void escreverVetorEmArquivo(const std::string& nomeDoArquivo, const std::vector<int>& vetor) {
    std::ofstream arquivo(nomeDoArquivo);
    if (!arquivo.is_open()) {
        std::cerr << "Erro ao criar o arquivo de saida: " << nomeDoArquivo << std::endl;
        return;
    }
    for (int numero : vetor) {
        arquivo << numero << "\n";
    }
}

int main() {
    std::string caminhoEntradaBase = "../../../Vetores/tamanho_100/";
    std::string caminhoSaidaBase = "../../Vetores ordenados/C++/Selection sort/";

    std::cout << "Iniciando processo de ordenacao em lote com Selection Sort (C++)..." << std::endl;

    for (int i = 1; i <= 100; ++i) {
        std::string nomeArquivoEntrada = caminhoEntradaBase + "vetor_100_" + std::to_string(i) + ".txt";
        std::string nomeArquivoSaida = caminhoSaidaBase + "vetor_100_" + std::to_string(i) + ".txt";
        std::vector<int> vetor = lerVetorDeArquivo(nomeArquivoEntrada);

        if (!vetor.empty()) {
            heapsort(vetor);
            escreverVetorEmArquivo(nomeArquivoSaida, vetor);
            std::cout << "Arquivo vetor_100_" << i << ".txt processado com sucesso." << std::endl;
        }
    }

    std::cout << "Processo concluido!" << std::endl;
    return 0;
}